File: src/api/languageProfiles.ts
import { browserInvoke } from "../lib/browser-backend";
import { invokeCommand, isTauri } from "../lib/tauri";
import {
  DEFAULT_LANGUAGE_PROFILE_SCOPE,
  DEFAULT_PROFILE_PREFERENCES,
  DEFAULT_PROFILE_PROCESSING_CONFIG,
  type ContentType,
  type DetectionEvidence,
  type LanguageProfile,
  type LanguageProfileAssociation,
  type LanguageProfileAssociationInput,
  type LanguageProfileCreate,
  type LanguageProfileExport,
  type LanguageProfileScope,
  type LanguageProfileSuggestion,
  type LanguageProfileSyncEnvelope,
  type LanguageProfileUpdate,
  type ProcessingConfig,
  type ProfileDeleteReport,
  type ResolvedLanguageProfileContext,
} from "../types/languageProfile";

type ProfileCall = <T>(command: string, args?: Record<string, unknown>) => Promise<T>;
const call: ProfileCall = (command, args) =>
  isTauri() ? invokeCommand(command, args) : browserInvoke(command, args);

function scopeArgs(scope?: Partial<LanguageProfileScope>): Record<string, string> {
  return {
    accountId: scope?.accountId || DEFAULT_LANGUAGE_PROFILE_SCOPE.accountId,
    workspaceId: scope?.workspaceId || DEFAULT_LANGUAGE_PROFILE_SCOPE.workspaceId,
  };
}

function createPayload(input: LanguageProfileCreate, scope?: Partial<LanguageProfileScope>) {
  return {
    input: {
      ...input,
      ...scopeArgs(scope),
      preferences: { ...DEFAULT_PROFILE_PREFERENCES, ...input.preferences },
      processingConfig: { ...DEFAULT_PROFILE_PROCESSING_CONFIG, ...input.processingConfig },
    },
    ...scopeArgs(scope),
  };
}

export function createLanguageProfile(
  input: LanguageProfileCreate,
  scope?: Partial<LanguageProfileScope>,
): Promise<LanguageProfile> {
  return call<LanguageProfile>("create_language_profile", createPayload(input, scope));
}

export function getLanguageProfiles(scope?: Partial<LanguageProfileScope>): Promise<LanguageProfile[]> {
  return call<LanguageProfile[]>("get_language_profiles", scopeArgs(scope));
}

export function getLanguageProfile(id: string, scope?: Partial<LanguageProfileScope>): Promise<LanguageProfile | null> {
  return call<LanguageProfile | null>("get_language_profile", { id, ...scopeArgs(scope) });
}

export function updateLanguageProfile(
  id: string,
  input: LanguageProfileUpdate,
  scope?: Partial<LanguageProfileScope>,
): Promise<LanguageProfile> {
  return call<LanguageProfile>("update_language_profile", { id, input, ...scopeArgs(scope) });
}

export function archiveLanguageProfile(id: string, scope?: Partial<LanguageProfileScope>): Promise<LanguageProfile> {
  return call<LanguageProfile>("archive_language_profile", { id, ...scopeArgs(scope) });
}

export function deleteLanguageProfile(id: string, scope?: Partial<LanguageProfileScope>): Promise<ProfileDeleteReport> {
  return call<ProfileDeleteReport>("delete_language_profile", { id, ...scopeArgs(scope) });
}

export function getActiveLanguageProfile(scope?: Partial<LanguageProfileScope>): Promise<LanguageProfile | null> {
  return call<LanguageProfile | null>("get_active_language_profile", scopeArgs(scope));
}

export function setActiveLanguageProfile(
  profileId: string | null,
  scope?: Partial<LanguageProfileScope>,
): Promise<LanguageProfile | null> {
  return call<LanguageProfile | null>("set_active_language_profile", { profileId, ...scopeArgs(scope) });
}

export function associateLanguageProfileContent(
  input: LanguageProfileAssociationInput,
  scope?: Partial<LanguageProfileScope>,
): Promise<LanguageProfileAssociation> {
  return call<LanguageProfileAssociation>("associate_language_profile_content", {
    input: { ...input, ...scopeArgs(scope) },
    ...scopeArgs(scope),
  });
}

export function getLanguageProfileAssociations(
  filters: { contentType?: ContentType; contentId?: string; profileId?: string } = {},
  scope?: Partial<LanguageProfileScope>,
): Promise<LanguageProfileAssociation[]> {
  return call<LanguageProfileAssociation[]>("get_language_profile_associations", {
    ...filters,
    ...scopeArgs(scope),
  });
}

export function resolveLanguageProfileContext(
  contentType: ContentType,
  contentId: string,
  explicitProfileId?: string | null,
  scope?: Partial<LanguageProfileScope>,
): Promise<ResolvedLanguageProfileContext | null> {
  return call<ResolvedLanguageProfileContext | null>("resolve_language_profile_context", {
    contentType,
    contentId,
    explicitProfileId: explicitProfileId || null,
    ...scopeArgs(scope),
  });
}

export function getLanguageProfileSuggestion(
  contentType: ContentType,
  contentId: string,
  evidence: DetectionEvidence,
  scope?: Partial<LanguageProfileScope>,
): Promise<LanguageProfileSuggestion | null> {
  return call<LanguageProfileSuggestion | null>("get_language_profile_suggestion", {
    contentType,
    contentId,
    evidence,
    ...scopeArgs(scope),
  });
}

export function dismissLanguageProfileSuggestion(
  profileId: string,
  contentType: ContentType,
  contentId: string,
  evidence?: DetectionEvidence,
  scope?: Partial<LanguageProfileScope>,
): Promise<LanguageProfileAssociation> {
  return call<LanguageProfileAssociation>("dismiss_language_profile_suggestion", {
    profileId,
    contentType,
    contentId,
    evidence: evidence || null,
    ...scopeArgs(scope),
  });
}

export function exportLanguageProfiles(scope?: Partial<LanguageProfileScope>): Promise<LanguageProfileExport> {
  return call<LanguageProfileExport>("export_language_profiles", scopeArgs(scope));
}

export function importLanguageProfiles(
  payload: LanguageProfileExport,
  conflict: "merge" | "replace" = "merge",
  scope?: Partial<LanguageProfileScope>,
): Promise<LanguageProfileExport> {
  return call<LanguageProfileExport>("import_language_profiles", { payload, conflict, ...scopeArgs(scope) });
}

export function serializeLanguageProfilesForSync(scope?: Partial<LanguageProfileScope>): Promise<LanguageProfileSyncEnvelope> {
  return call<LanguageProfileSyncEnvelope>("serialize_language_profiles_for_sync", scopeArgs(scope));
}

export function applyLanguageProfilesSync(
  payload: LanguageProfileSyncEnvelope,
  scope?: Partial<LanguageProfileScope>,
): Promise<LanguageProfileSyncEnvelope> {
  return call<LanguageProfileSyncEnvelope>("apply_language_profiles_sync", { payload, ...scopeArgs(scope) });
}

export type { ProcessingConfig };

File: src-tauri/src/commands/language_profiles.rs
//! Tauri boundary for learner profiles and explicit content associations.

use crate::database::Repository;
use crate::models::language_profile::{
    AssociationMode, ContentType, DetectionEvidence, LanguageProfile, LanguageProfileAssociation,
    LanguageProfileAssociationInput, LanguageProfileCreate, LanguageProfileExport,
    LanguageProfileSyncEnvelope, LanguageProfileUpdate, ProfileDeleteReport,
    ResolvedLanguageProfileContext,
};
use tauri::State;

#[tauri::command]
pub async fn create_language_profile(
    input: LanguageProfileCreate,
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<LanguageProfile, String> {
    repo.create_language_profile(input, account_id.as_deref(), workspace_id.as_deref())
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn get_language_profiles(
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<Vec<LanguageProfile>, String> {
    repo.get_language_profiles(account_id.as_deref(), workspace_id.as_deref())
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn get_language_profile(
    id: String,
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<Option<LanguageProfile>, String> {
    repo.get_language_profile(&id, account_id.as_deref(), workspace_id.as_deref())
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn update_language_profile(
    id: String,
    input: LanguageProfileUpdate,
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<LanguageProfile, String> {
    repo.update_language_profile(&id, input, account_id.as_deref(), workspace_id.as_deref())
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn archive_language_profile(
    id: String,
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<LanguageProfile, String> {
    repo.archive_language_profile(&id, account_id.as_deref(), workspace_id.as_deref())
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn delete_language_profile(
    id: String,
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<ProfileDeleteReport, String> {
    repo.delete_language_profile(&id, account_id.as_deref(), workspace_id.as_deref())
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn get_active_language_profile(
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<Option<LanguageProfile>, String> {
    repo.get_active_language_profile(account_id.as_deref(), workspace_id.as_deref())
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn set_active_language_profile(
    profile_id: Option<String>,
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<Option<LanguageProfile>, String> {
    repo.set_active_language_profile(
        profile_id.as_deref(),
        account_id.as_deref(),
        workspace_id.as_deref(),
    )
    .await
    .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn associate_language_profile_content(
    input: LanguageProfileAssociationInput,
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<LanguageProfileAssociation, String> {
    repo.upsert_language_profile_association(input, account_id.as_deref(), workspace_id.as_deref())
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn get_language_profile_associations(
    content_type: Option<String>,
    content_id: Option<String>,
    profile_id: Option<String>,
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<Vec<LanguageProfileAssociation>, String> {
    repo.get_language_profile_associations(
        content_type.as_deref(),
        content_id.as_deref(),
        profile_id.as_deref(),
        account_id.as_deref(),
        workspace_id.as_deref(),
    )
    .await
    .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn resolve_language_profile_context(
    content_type: String,
    content_id: String,
    explicit_profile_id: Option<String>,
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<Option<ResolvedLanguageProfileContext>, String> {
    repo.resolve_language_profile_context(
        &content_type,
        &content_id,
        explicit_profile_id.as_deref(),
        account_id.as_deref(),
        workspace_id.as_deref(),
    )
    .await
    .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn get_language_profile_suggestion(
    content_type: String,
    content_id: String,
    evidence: DetectionEvidence,
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<Option<crate::models::language_profile::LanguageProfileSuggestion>, String> {
    repo.get_language_profile_suggestion(
        &content_type,
        &content_id,
        evidence,
        account_id.as_deref(),
        workspace_id.as_deref(),
    )
    .await
    .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn dismiss_language_profile_suggestion(
    profile_id: String,
    content_type: ContentType,
    content_id: String,
    evidence: Option<DetectionEvidence>,
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<LanguageProfileAssociation, String> {
    repo.dismiss_language_profile_suggestion(
        &profile_id,
        content_type,
        &content_id,
        evidence,
        account_id.as_deref(),
        workspace_id.as_deref(),
    )
    .await
    .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn export_language_profiles(
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<LanguageProfileExport, String> {
    repo.export_language_profiles(account_id.as_deref(), workspace_id.as_deref())
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn import_language_profiles(
    payload: LanguageProfileExport,
    conflict: Option<String>,
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<LanguageProfileExport, String> {
    repo.import_language_profiles(
        payload,
        conflict.as_deref().unwrap_or("merge"),
        account_id.as_deref(),
        workspace_id.as_deref(),
    )
    .await
    .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn serialize_language_profiles_for_sync(
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<LanguageProfileSyncEnvelope, String> {
    repo.serialize_language_profiles_for_sync(account_id.as_deref(), workspace_id.as_deref())
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn apply_language_profiles_sync(
    payload: LanguageProfileSyncEnvelope,
    account_id: Option<String>,
    workspace_id: Option<String>,
    repo: State<'_, Repository>,
) -> Result<LanguageProfileSyncEnvelope, String> {
    repo.apply_language_profiles_sync(payload, account_id.as_deref(), workspace_id.as_deref())
        .await
        .map_err(|error| error.to_string())
}

// Keep the enum imported at this boundary as part of the command contract;
// this also makes its accepted JSON values obvious to generated bindings.
#[allow(dead_code)]
fn _association_mode_contract(mode: AssociationMode) -> AssociationMode {
    mode
}
